src/kernel/include/component.h

```diff
@@ -48,15 +48,18 @@ comp_activate(struct captbl *t, capid_t cap, capid_t capin, capid_t captbl_cap,
 	if (cos_cas((unsigned long *)&ptc->refcnt_flags, v, v + 1) != CAS_SUCCESS) return -ECASFAIL;
 
 	v = ctc->refcnt_flags;
-	if (v & CAP_MEM_FROZEN_FLAG) return -EINVAL;
+	if (v & CAP_MEM_FROZEN_FLAG) {
+		ret = -EINVAL;
+		goto undo_ptc;
+	}
 	if (cos_cas((unsigned long *)&ctc->refcnt_flags, v, v + 1) != CAS_SUCCESS) {
 		/* undo before return */
-		cos_faa((int *)&ptc->refcnt_flags, -1);
-		return -ECASFAIL;
+		ret = -ECASFAIL;
+		goto undo_ptc;
 	}
 	
 	compc = (struct cap_comp *)__cap_capactivate_pre(t, cap, capin, CAP_COMP, &ret);
-	if (!compc) return ret;
+	if (!compc) goto undo_ctc;
```

Comment:
I'd even use `cos_throw(ret, undo_ctc)` here even though it feels wasteful...it makes it more explicit where ret is getting updated.  I do wonder if the resulting `ret = ret;` will cause a compiler warning, in which case we should not do it.